conformance/utils/weight/weight.go

```diff
@@ -0,0 +1,179 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package weight
+
+import (
+	"cmp"
+	"crypto/rand"
+	"errors"
+	"fmt"
+	"math"
+	"math/big"
+	"slices"
+	"strings"
+	"sync"
+	"time"
+
+	"golang.org/x/sync/errgroup"
+)
+
+// RequestSender defines an interface for sending requests (HTTP, gRPC, or mesh)
+type RequestSender interface {
+	SendRequest() (podName string, err error)
+}
+
+// TestWeightedDistribution tests that requests are distributed according to expected weights
+func TestWeightedDistribution(sender RequestSender, expectedWeights map[string]float64) error {
+	const (
+		concurrentRequests  = 10
+		tolerancePercentage = 0.05
+		totalRequests       = 500.0
+	)
+
+	var (
+		g         errgroup.Group
+		seenMutex sync.Mutex
+		seen      = make(map[string]float64, len(expectedWeights))
+	)
+
+	g.SetLimit(concurrentRequests)
+	for i := 0.0; i < totalRequests; i++ {
+		g.Go(func() error {
+			podName, err := sender.SendRequest()
+			if err != nil {
+				return err
+			}
+
+			seenMutex.Lock()
+			defer seenMutex.Unlock()
+
+			for expectedBackend := range expectedWeights {
+				if strings.HasPrefix(podName, expectedBackend) {
+					seen[expectedBackend]++
+					return nil
+				}
+			}
+
+			return fmt.Errorf("request was handled by an unexpected pod %q", podName)
+		})
+	}
+
+	if err := g.Wait(); err != nil {
+		return fmt.Errorf("error while sending requests: %w", err)
+	}
+
+	// Count how many backends should receive traffic (weight > 0)
+	expectedActiveBackends := 0
+	for _, weight := range expectedWeights {
+		if weight > 0.0 {
+			expectedActiveBackends++
+		}
+	}
+
+	var errs []error
+	if len(seen) != expectedActiveBackends {
+		errs = append(errs, fmt.Errorf("expected %d backends to receive traffic, but got %d", expectedActiveBackends, len(seen)))
+	}
+
+	for wantBackend, wantPercent := range expectedWeights {
+		gotCount, ok := seen[wantBackend]
+
+		if !ok && wantPercent != 0.0 {
+			errs = append(errs, fmt.Errorf("expect traffic to hit backend %q - but none was received", wantBackend))
+			continue
+		}
+
+		gotPercent := gotCount / totalRequests
+
+		if math.Abs(gotPercent-wantPercent) > tolerancePercentage {
+			errs = append(errs, fmt.Errorf("backend %q weighted traffic of %v not within tolerance %v (+/-%f)",
+				wantBackend,
+				gotPercent,
+				wantPercent,
+				tolerancePercentage,
+			))
+		}
+	}
+
+	slices.SortFunc(errs, func(a, b error) int {
+		return cmp.Compare(a.Error(), b.Error())
+	})
+	return errors.Join(errs...)
+}
+
+// Entropy utilities
+
+// randomNumber generates a random number between 0 and limit-1
+func randomNumber(limit int64) (*int64, error) {
+	number, err := rand.Int(rand.Reader, big.NewInt(limit))
+	if err != nil {
+		return nil, err
+	}
+	n := number.Int64()
+	return &n, nil
+}
+
+// AddDelay adds a random delay up to the specified limit in milliseconds
+func AddDelay(limit int64) error {
```

Comment:
```suggestion
// addRandomDelay adds a random delay up to the specified limit in milliseconds
func addRandomDelay(limit int64) error {
```
I think addRandomDelay is a more descriptive name and also I think it should be a private function.